src/net_processing.cpp

```diff
@@ -254,6 +271,71 @@ struct CNodeState {
     //! Time of last new block announcement
     int64_t m_last_block_announcement;
 
+    /*
+     * State associated with transaction download.
+     *
+     * Tx download algorithm:
+     *
+     *   When inv comes in, queue up (process_time, txid) inside the peer's
+     *   CNodeState (m_tx_process_time) as long as m_tx_announced for the peer
+     *   isn't too big (MAX_PEER_TX_ANNOUNCEMENTS).
+     *
+     *   The process_time for a transaction is set to nNow for outbound peers,
+     *   nNow + 2 seconds for inbound peers. This is the time at which we'll
+     *   consider trying to request the transaction from the peer in
+     *   SendMessages(). The delay for inbound peers is to allow outbound peers
+     *   a chance to announce before we request from inbound peers, to prevent
+     *   an adversary from using inbound connections to blind us to a
+     *   transaction (InvBlock).
+     *
+     *   When we call SendMessages() for a given peer,
+     *   we will loop over the transactions in m_tx_process_time, looking
+     *   at the transactions whose process_time <= nNow. We'll request each
+     *   such transaction that we don't have already and that hasn't been
+     *   requested from another peer recently, up until we hit the
+     *   MAX_PEER_TX_IN_FLIGHT limit for the peer. Then we'll update
+     *   g_already_asked_for for each requested txid, storing the time of the
+     *   GETDATA request. We use g_already_asked_for to coordinate transaction
+     *   requests amongst our peers.
+     *
+     *   For transactions that we still need but we have already recently
+     *   requested from some other peer, we'll reinsert (process_time, txid)
+     *   back into the peer's m_tx_process_time at the point in the future at
+     *   which the most recent GETDATA request would time out (ie
+     *   GETDATA_TX_INTERVAL + the request time stored in g_already_asked_for).
+     *   We add an additional delay for inbound peers, again to prefer
+     *   attempting download from outbound peers first.
+     *   We also add an extra small random delay up to 2 seconds
+     *   to avoid biasing some peers over others. (e.g., due to fixed ordering
+     *   of peer processing in ThreadMessageHandler).
+     *
+     *   When we receive a transaction from a peer, we remove the txid from the
+     *   peer's m_tx_in_flight set and from their recently announced set
+     *   (m_tx_announced).  We also clear g_already_asked_for for that entry, so
+     *   that if somehow the transaction is not accepted but also not added to
+     *   the reject filter, then we will eventually redownload from other
+     *   peers.
+     *
+     *   DASH: For Dash, this does not only handles TXs but also all Dash specific objects
```

Comment:
```suggestion
     *   DASH: For Dash, this not only handles TXs but also all Dash specific objects
```

Reply:
I actually agree, but I'm super worried that this will cause many merge conflicts in the future.

Reply:
I only see changes from https://github.com/bitcoin/bitcoin/pull/16197 in related code in btc `master`, should be trivial to backport (might be a good idea to do it here?).

Reply:
Applied c17f7dfff8d2b33a8de4eb6e77ccdca7e39fd338. I tried to backport bitcoin#16197 and stumbled across too much missing stuff in the test framework. Better to do that backport in-order.

Reply:
Ahhh, I just realized that I can also just apply the single commit fa883ab35ad2d4328e35b1e855d0833740a6b910 from that PR, so that the new tests are not included. Later when we do the whole PR backport, we'll see that it was partly backported already.

Reply:
Done. I also had to backport bitcoin#16046 for this to work.